feather/common/src/block.rs

```diff
@@ -0,0 +1,595 @@
+        {
+            target2
+        } else {
+            return None;
+        }
+    };
+    door_hinge(&mut block, target, &placement.cursor_position, world);
+    let place_top = match top_half(block) {
+        Some(_) => {
+            if !world
+                .block_at(target.up().try_into().unwrap())?
+                .is_replaceable()
+            {
+                // Short circuit if upper block is > 256
+                return None;
+            }
+            true
+        }
+        None => false,
+    };
+    let place_head = match bed_head(block) {
+        Some(_) => {
+            if !world
+                .adjacent_block_cardinal(target, block.facing_cardinal()?)?
+                .is_replaceable()
+            {
+                return None;
+            }
+            true
+        }
+        None => false,
+    };
+    // This works but there is a discrepancy between when the place block event is fired and getting the entity location.
+    // that makes it possible to place a block at the right exact moment and have the server believe it wasn't blocked.
+    if chunk_entities
+        .entities_in_chunk(target.chunk())
+        .iter()
+        .any(|_entity| {
+            let entity_position = ecs.get::<Position>(*_entity).unwrap();
+            let entity_kind = *ecs.get::<EntityKind>(*_entity).unwrap();
+            let block_rect: Rect3<f64, f64> = vek::Rect3 {
+                x: target.x.into(),
+                y: target.y.into(),
+                z: target.z.into(),
+                w: 1.0,
+                h: 1.0,
+                d: 1.0,
+            };
+
+            let mut entity_rect = entity_kind.bounding_box().into_rect3();
+            entity_rect.x = entity_position.x - (entity_rect.w / 2.0);
+            entity_rect.y = entity_position.y;
+            entity_rect.z = entity_position.z - (entity_rect.d / 2.0);
+
+            block_rect.collides_with_rect3(entity_rect)
+        })
+    {
+        return None;
+    }
+    if !world.check_block_stability(block, target, light_level)? {
+        return None;
+    }
+    if place_top {
+        world.set_block_at(target.try_into().unwrap(), block);
+        world.set_block_at(
+            target.up().try_into().unwrap(),
+            block.with_half_upper_lower(HalfUpperLower::Upper),
+        );
+        Some(vec![
+            BlockChangeEvent::single(target.try_into().unwrap()),
+            BlockChangeEvent::single(target.up().try_into().unwrap()),
+        ])
+    } else if place_head {
+        let face = match block.facing_cardinal()? {
+            FacingCardinal::North => BlockFace::North,
+            FacingCardinal::South => BlockFace::South,
+            FacingCardinal::West => BlockFace::West,
+            FacingCardinal::East => BlockFace::East,
+        };
+        if !world.check_block_stability(block, target.adjacent(face), light_level)? {
+            return None;
+        }
```

Comment:
Does the bed really need a stability check? If you just want to make sure it's stable, you should also add this check to place_top (it would be a bit trickier because you'll need to set the lower half of the door, then check and undo it if it's not stable)

Reply:
You are right, it doesn't. This shouldn't a problem with place_top, I don't know of any double block that has extra requirements for its top half.